modules/p2/ding_interests/ding_interests.module

```diff
@@ -0,0 +1,652 @@
+
+  $items['admin/config/ding/interests'] = array(
+    'title' => 'Ding interests settings',
+    'description' => 'Setup and configure ding interests.',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('ding_interests_admin_form'),
+    'access arguments' => array('administer personalisation'),
+    'file' => 'ding_interests.admin.inc',
+  );
+  $items['ding/interests/%ctools_js/%/%user/%'] = array(
+    'title' => 'Ding interests add to list',
+    'page callback' => 'ding_interests_add_to_list_callback',
+    'page arguments' => array(2, 3, 4, 5),
+    'access arguments' => array('use personalisation'),
+  );
+  $items['ding/interests/remove/%ctools_js/%taxonomy_term/%user'] = array(
+    'title' => 'Ding interests remove from list',
+    'page callback' => 'ding_interests_remove_from_list_callback',
+    'page arguments' => array(2, 3, 4, 5),
+    'access arguments' => array('use personalisation'),
+  );
+  $items['ding/interests/overview'] = array(
+    'title' => 'List of all recommended subjects',
+    'page callback' => 'ding_interests_all_recommended_subjects',
+    'access arguments' => array('use personalisation'),
+  );
+
+  return $items;
+}
+
+/**
+ * Implements hook_entity_info_alter().
+ */
+function ding_interests_entity_info_alter(&$entity_info) {
+  // Taxonomy teasers are used for recommending interests.
+  if (!isset($entity_info['taxonomy_term']['view modes']['teaser'])) {
+    $entity_info['taxonomy_term']['view modes']['teaser'] = array(
+      'label' => t('Teaser'),
+      'custom settings' => TRUE,
+    );
+  }
+}
+
+/**
+ * Synchronize with openlist.
+ *
+ * Create missing lists and update the historical loans list.
+ */
+function ding_interests_ding_list_operations() {
+  return array(
+    DING_INTERESTS_LIST => array(
+      DING_LIST_OPERATION_DETACH => TRUE,
+      DING_LIST_OPERATION_SORT => TRUE,
+      DING_LIST_OPERATION_AUTO_CREATE => TRUE,
+      '_data' => array(
+        'default_title' => t('Interests'),
+        'description' => t('Follow your favorite topics and recieve notifications when new materials arrive. <a href="/ding/interests/overview">Check out recommended topics here</a>'),
+        'yes' => t('Interests'),
+        'no' => t('Topics I have no interrest in.'),
+      ),
+    ),
+  );
+}
+
+/**
+ * Implements hook_ding_interaction_view().
+ */
+function ding_interests_ding_interaction_view() {
+  if (user_is_logged_in()) {
+    
+    // UI for following interests is ajax time.
+    ctools_include('ajax');
+    $interests_list = ding_list_get_list(DING_INTERESTS_LIST);
+
+    // Try to find the correct status tid.
+    $status_voc = taxonomy_vocabulary_machine_name_load('interest_status');
+    $status_terms = taxonomy_get_tree($status_voc->vid);
+
+    foreach ($status_terms as $status_term) {
+      if ($status_term->tid == variable_get('ding_interests_selected_tid')) {
+        $status_tid = $status_term->tid;
+      }
+    }
+
+    $result = db_query("SELECT
```

Comment:
An explanion of what this 18 LOC query does and why would be highly appreciated.